C++: add missing implementations for client advertised schema/encoding

paulsohn opened the pull request:
### Changelog
<!-- Write a one-sentence summary of the user-impacting change (API, UI/UX, performance, etc) that could appear in a changelog. Write "None" if there is no user-facing change -->

C++ server/client: add missing implementations for client advertised schema / schema encoding

### Docs

<!-- Link to a Docs PR, tracking ticket in Linear, OR write "None" if no documentation changes are needed. -->

* Issue #875 (Initial discussions are here)
* Liner Ticket FG-9447 (though I can't access to it)

### Description

<!-- Describe the problem, what has changed, and motivation behind those changes. Pretend you are advocating for this change and the reader is skeptical. -->

Although foxglove ws protocol allow client to advertise complete schema definition string as well as schema name, current C++ implementation lacks such fields. Particularly:
* `ClientAdvertisement` struct lacks `schemaEncoding` field,
* `ClientAdvertisement` struct has `schema` field, but it is never populated (websocket server does not read schema or schema encoding to json)
* websocket client is does not serialize schema or schema encoding to json

This PR resolves above problems, and additionally since this is breaking change from unimplemented to implemented, types are decided as following:
* `schema` field type is changed from `std::vector<uint8_t>` to `std::optional<std::string>`, as schema fields are passed in as json quoted string, and it's optional
* `schemaEncoding` field type is determined as `std::optional<std::string>`, the same reason for `schema`.

<!-- In addition to unit tests, describe any manual testing you did to validate this change. -->

<!-- If necessary, link relevant Linear or Github issues. Use `Fixes: foxglove/repo#1234` to auto-close the Github issue or Fixes: FG-### for Linear isses. -->

Fixes: foxglove/ws-protocol#875
Fixes: FG-9447

defunctzombie commented:
@paulsohn Thanks for the proposed fix here - let us know when this is ready for review!

paulsohn commented:
@defunctzombie Sure.
It is a draft because I got no tests but I don't think further changes might be needed, so feel free to test this as it is (I will do my tests in next week)

paulsohn commented:
Fixed error.

Added some test in separate branch ( https://github.com/paulsohn/ws-protocol/tree/test-pr876 ).
After build, execute `make example_server_client_pub_json` and `make example_client_pub_json` on separate shell in this order.

paulsohn commented:
Are you currently reviewing this? If you don't mind, let me know the progress so far

By the way I just thought, it might affect ros-foxglove-bridge as well?

https://github.com/foxglove/ros-foxglove-bridge/blob/main/ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

achim-k commented on cpp/foxglove-websocket/include/foxglove/websocket/websocket_server.hpp:
```diff
@@ -1340,6 +1340,14 @@ void Server<ServerConfiguration>::handleAdvertise(const nlohmann::json& payload,
     advertisement.encoding = chan.at("encoding").get<std::string>();
     advertisement.schemaName = chan.at("schemaName").get<std::string>();
 
+    advertisement.schema = chan.find("schema") == chan.end()
+                             ? std::nullopt
+                             : std::optional<std::string>(chan["schema"].get<std::string>());
+    advertisement.schemaEncoding =
+      chan.find("schemaEncoding") == chan.end()
+        ? std::nullopt
+        : std::optional<std::string>(chan["schemaEncoding"].get<std::string>());
+
```
I believe the following should also work which avoids a second lookup of the json property

```suggestion
    if (const auto it = chan.find("schema"); it != chan.end()) {
      advertisement.schema = it->get<std::string>();
    }
    if (const auto it = chan.find("schemaEncoding"); it != chan.end()) {
      advertisement.schemaEncoding = it->get<std::string>();
    }
```

paulsohn replied:
Some of the previous code accessed fields like this and others used iterators just as you're suggesting, so I just picked one randomly. Same for the another suggestion.

Nonetheless, I agree with both of your points. Should I make a new commit or will you?

achim-k replied:
Ah ok, was not aware of this.

achim-k commented on cpp/foxglove-websocket/include/foxglove/websocket/websocket_client.hpp:
```diff
@@ -23,6 +23,12 @@ inline void to_json(nlohmann::json& j, const ClientAdvertisement& p) {
                      {"topic", p.topic},
                      {"encoding", p.encoding},
                      {"schemaName", p.schemaName}};
+  if (p.schema) {
+    j["schema"] = *p.schema;
+  }
+  if (p.schemaEncoding) {
+    j["schemaEncoding"] = *p.schemaEncoding;
+  }
```
```suggestion
  if (p.schema) {
    j["schema"] = p.schema.value();
  }
  if (p.schemaEncoding) {
    j["schemaEncoding"] = p.schemaEncoding.value();
  }
```

To be consistent with the rest of the codebase

jtbandes commented:
Thank you!
